src/middlewared/middlewared/plugins/failover_/datastore.py

```diff
@@ -0,0 +1,125 @@
+import os
+import time
+
+from middlewared.service import CallError, Service
+from middlewared.plugins.config import FREENAS_DATABASE
+from middlewared.plugins.datastore.connection import thread_pool
+from middlewared.utils.threading import start_daemon_thread, set_thread_name
+
+FREENAS_DATABASE_REPLICATED = f'{FREENAS_DATABASE}.replicated'
+
+
+class FailoverDatastoreService(Service):
+
+    class Config:
+        namespace = 'failover.datastore'
+        private = True
+
+    thread_pool = thread_pool
+
+    def sql(self, data, sql, params):
+        if self.middleware.call_sync('system.version') != data['version']:
+            return
+
+        self.middleware.call_sync('datastore.execute', sql, params)
+
+    failure = False
+
+    def is_failure(self):
+        return self.failure
+
+    def set_failure(self):
+        self.failure = True
+        try:
+            self.send()
+        except Exception as e:
+            self.middleware.logger.warning(
```

Comment:
```suggestion
            self.logger.warning(
```

Also get rid of `e` and use `exc_info=True`